### ISSUE TITLE
UdpSocket-related deadlock when dropping runtime

### ISSUE THREAD

**Issue by dfoxfranke** [tokio-rs/tokio#2061]
This bug appears to be distinct from #2046.

## Version
```
deadlock v0.1.0 (/home/dafranke/deadlock)
└── tokio v0.2.6
    ├── bytes v0.5.3
    ├── fnv v1.0.6
    ├── futures-core v0.3.1
    ├── iovec v0.1.4
    │   └── libc v0.2.66
    ├── lazy_static v1.4.0
    ├── libc v0.2.66 (*)
    ├── memchr v2.2.1
    ├── mio v0.6.21
    │   ├── cfg-if v0.1.10
    │   ├── iovec v0.1.4 (*)
    │   ├── libc v0.2.66 (*)
    │   ├── log v0.4.8
    │   │   └── cfg-if v0.1.10 (*)
    │   ├── net2 v0.2.33
    │   │   ├── cfg-if v0.1.10 (*)
    │   │   └── libc v0.2.66 (*)
    │   └── slab v0.4.2
    ├── mio-uds v0.6.7
    │   ├── iovec v0.1.4 (*)
    │   ├── libc v0.2.66 (*)
    │   └── mio v0.6.21 (*)
    ├── num_cpus v1.11.1
    │   └── libc v0.2.66 (*)
    ├── pin-project-lite v0.1.2
    ├── signal-hook-registry v1.2.0
    │   ├── arc-swap v0.4.4
    │   └── libc v0.2.66 (*)
    ├── slab v0.4.2 (*)
    └── tokio-macros v0.2.1
        ├── quote v1.0.2
        │   └── proc-macro2 v1.0.7
        │       └── unicode-xid v0.2.0
        └── syn v1.0.13
            ├── proc-macro2 v1.0.7 (*)
            ├── quote v1.0.2 (*)
            └── unicode-xid v0.2.0 (*)
```
## Platform

`Linux chocorua 4.19.91 #1-NixOS SMP Sat Dec 21 09:57:45 UTC 2019 x86_64 GNU/Linux`

## Description

The following program deadlocks:

```rust
use tokio::runtime::Runtime;
use tokio::net::UdpSocket;
use std::net::Ipv6Addr;
use std::time::Duration;

fn main() {
    for _ in 1..100 {
        let mut runtime = Runtime::new().unwrap();

        runtime.block_on(async {
            let socket = UdpSocket::bind((Ipv6Addr::LOCALHOST, 0)).await.unwrap();
            let addr = socket.local_addr().unwrap();
            let (mut recv_half, mut send_half) = socket.split();

            tokio::spawn(async move {
                let mut buf = [0];
                loop {
                    recv_half.recv_from(&mut buf).await.unwrap();
                    std::thread::sleep(Duration::from_millis(2));
                }
            });

            tokio::spawn(async move {
                let buf = [0];
                loop {
                    send_half.send_to(&buf, &addr).await.unwrap();
                    tokio::time::delay_for(Duration::from_millis(1)).await;
                }
            });

            tokio::time::delay_for(Duration::from_millis(5)).await;
        });
    }
}
```

Backtrace from lldb:

```
(lldb) thread list
Process 25794 stopped
* thread #1: tid = 25794, 0x00007f6f38c6eee2 libpthread.so.0`__pthread_cond_wait + 274, name = 'deadlock', stop reason = signal SIGSTOP
  thread #2: tid = 25843, 0x00007f6f38988567 libc.so.6`epoll_wait + 103, name = 'tokio-runtime-w', stop reason = signal SIGSTOP
(lldb) bt
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
* thread #1, name = 'deadlock', stop reason = signal SIGSTOP
  * frame #0: 0x00007f6f38c6eee2 libpthread.so.0`__pthread_cond_wait + 274
    frame #1: 0x00005571b2f57123 deadlock`std::sys::unix::condvar::Condvar::wait::h931978fe7d80a479(self=0x00005571b3984a30, mutex=0x00005571b3984a00) at condvar.rs:69
    frame #2: 0x00005571b2f570d7 deadlock`std::sys_common::condvar::Condvar::wait::h45f6267a90a58bb4(self=0x00005571b3984a30, mutex=0x00005571b3984a00) at condvar.rs:41
    frame #3: 0x00005571b2f529c1 deadlock`std::sync::condvar::Condvar::wait::h794ed8cfae369a29(self=0x00005571b3984a98, guard=MutexGuard<()> @ 0x00007ffcc9471040) at condvar.rs:204
    frame #4: 0x00005571b2eb8c9b deadlock`tokio::park::thread::Inner::park::hf12334e567af9fc0(self=0x00005571b3984a80) at thread.rs:114
    frame #5: 0x00005571b2eb9a21 deadlock`_$LT$tokio..park..thread..CachedParkThread$u20$as$u20$tokio..park..Park$GT$::park::_$u7b$$u7b$closure$u7d$$u7d$::hb6bb8407a5f45fc2((null)=closure-0 @ 0x00007ffcc9471368, park_thread=0x00007f6f388907f8) at thread.rs:253
    frame #6: 0x00005571b2eb9970 deadlock`tokio::park::thread::CachedParkThread::with_current::_$u7b$$u7b$closure$u7d$$u7d$::hac47a7515ead7d31(inner=0x00007f6f388907f8) at thread.rs:240
    frame #7: 0x00005571b2eb5078 deadlock`std::thread::local::LocalKey$LT$T$GT$::try_with::hf648fbdcb4a8d443(self=0x00005571b2feb4f8, f=(__0 = tokio::park::thread::{{impl}}::park::closure-0 @ 0x00007ffcc94713d0)) at local.rs:262
    frame #8: 0x00005571b2eb30d4 deadlock`std::thread::local::LocalKey$LT$T$GT$::with::hc4995a53e04d5151(self=0x00005571b2feb4f8, f=(__0 = tokio::park::thread::{{impl}}::park::closure-0 @ 0x00007ffcc9471450)) at local.rs:239
    frame #9: 0x00005571b2eb98e5 deadlock`tokio::park::thread::CachedParkThread::with_current::h01fee34a741a4767(self=0x00007ffcc9471500, f=closure-0 @ 0x00007ffcc9471468) at thread.rs:240
    frame #10: 0x00005571b2eb99e3 deadlock`_$LT$tokio..park..thread..CachedParkThread$u20$as$u20$tokio..park..Park$GT$::park::h37270cd724a0ec34(self=0x00007ffcc9471500) at thread.rs:253
    frame #11: 0x00005571b2eb9f29 deadlock`tokio::runtime::enter::Enter::block_on::h03a6a290ce1b1234(self=0x00007ffcc9471588, f=0x00007ffcc9471758) at enter.rs:103
    frame #12: 0x00005571b2eb21f8 deadlock`tokio::runtime::blocking::shutdown::Receiver::wait::h065667ad6b90613b(self=0x00007ffcc9471758) at shutdown.rs:42
    frame #13: 0x00005571b2f03175 deadlock`_$LT$tokio..runtime..blocking..pool..BlockingPool$u20$as$u20$core..ops..drop..Drop$GT$::drop::h6dd880fa6f3c9cc9(self=0x00007ffcc9471750) at pool.rs:151
    frame #14: 0x00005571b2e96ca5 deadlock`core::ptr::real_drop_in_place::h925151c8c47a14ef((null)=0x00007ffcc9471750) at mod.rs:175
    frame #15: 0x00005571b2e9635e deadlock`core::ptr::real_drop_in_place::h4abaf3542119b019((null)=0x00007ffcc94716b0) at mod.rs:175
    frame #16: 0x00005571b2e8c6da deadlock`deadlock::main::hb7b0e92d6ffd824a at main.rs:33
    frame #17: 0x00005571b2e98040 deadlock`std::rt::lang_start::_$u7b$$u7b$closure$u7d$$u7d$::h72fe98da7ae2f61f at rt.rs:61
    frame #18: 0x00005571b2f72e53 deadlock`std::panicking::try::do_call::h631c6408dfccc6f5 [inlined] std::rt::lang_start_internal::_$u7b$$u7b$closure$u7d$$u7d$::h6ea535ec5c50fc3e at rt.rs:48
    frame #19: 0x00005571b2f72e47 deadlock`std::panicking::try::do_call::h631c6408dfccc6f5 at panicking.rs:287
    frame #20: 0x00005571b2f75e6a deadlock`__rust_maybe_catch_panic at lib.rs:78
    frame #21: 0x00005571b2f7390d deadlock`std::rt::lang_start_internal::h3bdc4c7d98181bf9 [inlined] std::panicking::try::hab539b2d1255d635 at panicking.rs:265
    frame #22: 0x00005571b2f738cf deadlock`std::rt::lang_start_internal::h3bdc4c7d98181bf9 [inlined] std::panic::catch_unwind::hd5e0a26424bd7f34 at panic.rs:396
    frame #23: 0x00005571b2f738cf deadlock`std::rt::lang_start_internal::h3bdc4c7d98181bf9 at rt.rs:47
    frame #24: 0x00005571b2e98019 deadlock`std::rt::lang_start::hd9a66c4ddf661574(main=(deadlock`deadlock::main::hb7b0e92d6ffd824a at main.rs:6), argc=1, argv=0x00007ffcc9471c08) at rt.rs:61
    frame #25: 0x00005571b2e8c71a deadlock`main + 42
    frame #26: 0x00007f6f388b7b8e libc.so.6`__libc_start_main + 238
    frame #27: 0x00005571b2e7916a deadlock`_start at start.S:120
(lldb) t 2
* thread #2, name = 'tokio-runtime-w', stop reason = signal SIGSTOP
    frame #0: 0x00007f6f38988567 libc.so.6`epoll_wait + 103
libc.so.6`epoll_wait:
->  0x7f6f38988567 <+103>: cmpq   $-0x1000, %rax            ; imm = 0xF000 
    0x7f6f3898856d <+109>: ja     0x7f6f3898858a            ; <+138>
    0x7f6f3898856f <+111>: movl   %r8d, %edi
    0x7f6f38988572 <+114>: movl   %eax, 0xc(%rsp)
(lldb) bt
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
* thread #2, name = 'tokio-runtime-w', stop reason = signal SIGSTOP
  * frame #0: 0x00007f6f38988567 libc.so.6`epoll_wait + 103
    frame #1: 0x00005571b2f59fe7 deadlock`mio::sys::unix::epoll::Selector::select::hce185c19994df7d9(self=0x00005571b397d080, evts=0x00005571b397de68, awakener=(__0 = 18446744073709551615), timeout=<unavailable>) at epoll.rs:72
    frame #2: 0x00005571b2f4d87b deadlock`mio::poll::Poll::poll2::hb2b59bcae2af14dc(self=0x00005571b397d080, events=0x00005571b397de68, timeout=Option<core::time::Duration> @ 0x00007f6f38283950, interruptible=false) at poll.rs:1178
    frame #3: 0x00005571b2f4d555 deadlock`mio::poll::Poll::poll1::h73e4b57e7bd070ce(self=0x00005571b397d080, events=0x00005571b397de68, timeout=Option<core::time::Duration> @ 0x00007f6f382839e0, interruptible=false) at poll.rs:1139
    frame #4: 0x00005571b2f4cde7 deadlock`mio::poll::Poll::poll::h0d5bc6d71d78d97b(self=0x00005571b397d080, events=0x00005571b397de68, timeout=<unavailable>) at poll.rs:1010
    frame #5: 0x00005571b2eda2d9 deadlock`tokio::io::driver::Driver::turn::hfe114e73c873acbb(self=0x00005571b397de68, max_wait=Option<core::time::Duration> @ 0x00007f6f38283b88) at mod.rs:147
    frame #6: 0x00005571b2edac88 deadlock`_$LT$tokio..io..driver..Driver$u20$as$u20$tokio..park..Park$GT$::park::he01a19793054ddb3(self=0x00005571b397de68) at mod.rs:216
    frame #7: 0x00005571b2ee7419 deadlock`_$LT$tokio..park..either..Either$LT$A$C$B$GT$$u20$as$u20$tokio..park..Park$GT$::park::ha9f9d4f7cff5b0b8(self=0x00005571b397de60) at either.rs:28
    frame #8: 0x00005571b2ed13f1 deadlock`_$LT$tokio..time..driver..Driver$LT$T$GT$$u20$as$u20$tokio..park..Park$GT$::park::h3ef0060fe91a7bc0(self=0x00005571b397de38) at mod.rs:262
    frame #9: 0x00005571b2ee7378 deadlock`_$LT$tokio..park..either..Either$LT$A$C$B$GT$$u20$as$u20$tokio..park..Park$GT$::park::h9ce8baca73fa5a80(self=0x00005571b397de30) at either.rs:28
    frame #10: 0x00005571b2edccf2 deadlock`tokio::runtime::park::Inner::park_driver::h75350090536f6b36(self=0x00005571b397e080, driver=0x00005571b397de30) at park.rs:186
    frame #11: 0x00005571b2edc523 deadlock`tokio::runtime::park::Inner::park::h20609f42a67582ef(self=0x00005571b397e080) at park.rs:128
    frame #12: 0x00005571b2edbfa3 deadlock`_$LT$tokio..runtime..park..Parker$u20$as$u20$tokio..park..Park$GT$::park::hbd86465f394d1bf4(self=0x00005571b397cb50) at park.rs:91
    frame #13: 0x00005571b2efc3fd deadlock`tokio::runtime::thread_pool::worker::GenerationGuard::shutdown::h2af626c053c09fa1(self=0x00007f6f38284280) at worker.rs:573
    frame #14: 0x00005571b2efae15 deadlock`tokio::runtime::thread_pool::worker::GenerationGuard::run::hc3749ea7c6c1437e(self=GenerationGuard @ 0x00007f6f38284270) at worker.rs:296
    frame #15: 0x00005571b2efa84f deadlock`tokio::runtime::thread_pool::worker::Worker::run::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::hf03183653399d4ca(ob=0x00007f6f38286650) at worker.rs:175
    frame #16: 0x00005571b2eb44f0 deadlock`std::thread::local::LocalKey$LT$T$GT$::try_with::ha1744ccaacc49b84(self=0x00005571b2fec1e8, f=closure-0 @ 0x00007f6f38284420) at local.rs:262
    frame #17: 0x00005571b2eb2da3 deadlock`std::thread::local::LocalKey$LT$T$GT$::with::h6f3e9e5be6e6bdae(self=0x00005571b2fec1e8, f=<unavailable>) at local.rs:239
    frame #18: 0x00005571b2efa8dd deadlock`tokio::runtime::thread_pool::worker::Worker::run::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h54fe0bcf0e93c43d at worker.rs:151
    frame #19: 0x00005571b2f033b3 deadlock`tokio::runtime::blocking::pool::Spawner::enter::_$u7b$$u7b$closure$u7d$$u7d$::h3316bb60beaa3cd8(cell=0x00007f6f38286560) at pool.rs:194
    frame #20: 0x00005571b2eb3fb0 deadlock`std::thread::local::LocalKey$LT$T$GT$::try_with::h7135afa89bd08ce0(self=0x00005571b2fec3d8, f=closure-0 @ 0x00007f6f382845f8) at local.rs:262
    frame #21: 0x00005571b2eb2b2c deadlock`std::thread::local::LocalKey$LT$T$GT$::with::h06253d134aa908b7(self=0x00005571b2fec3d8, f=<unavailable>) at local.rs:239
    frame #22: 0x00005571b2f0326c deadlock`tokio::runtime::blocking::pool::Spawner::enter::h0d76fd15264b49fc(self=0x00007f6f38284b50, f=<unavailable>) at pool.rs:179
    frame #23: 0x00005571b2efa91c deadlock`tokio::runtime::thread_pool::worker::Worker::run::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h852aec8b5ac66155 at worker.rs:150
    frame #24: 0x00005571b2e9de87 deadlock`tokio::runtime::global::with_state::_$u7b$$u7b$closure$u7d$$u7d$::heff94beb2ede511b(cell=0x00007f6f38286540) at global.rs:100
    frame #25: 0x00005571b2eb48f0 deadlock`std::thread::local::LocalKey$LT$T$GT$::try_with::hd4941bbb79263aa1(self=0x00005571b2feacd0, f=closure-0 @ 0x00007f6f382847f8) at local.rs:262
    frame #26: 0x00005571b2eb2d2c deadlock`std::thread::local::LocalKey$LT$T$GT$::with::h4e21d5129fb9b058(self=0x00005571b2feacd0, f=<unavailable>) at local.rs:239
    frame #27: 0x00005571b2e9dd64 deadlock`tokio::runtime::global::with_state::h6da15b269993f75a(state=State @ 0x00007f6f38284828, f=<unavailable>) at global.rs:83
    frame #28: 0x00005571b2e9e0db deadlock`tokio::runtime::global::with_thread_pool::h26c3e1f2d0d1589d(thread_pool=0x00007f6f38284b70, f=<unavailable>) at global.rs:75
    frame #29: 0x00005571b2efa97c deadlock`tokio::runtime::thread_pool::worker::Worker::run::_$u7b$$u7b$closure$u7d$$u7d$::h43b658d54639ef76 at worker.rs:149
    frame #30: 0x00005571b2e9e28e deadlock`tokio::runtime::thread_pool::current::set::_$u7b$$u7b$closure$u7d$$u7d$::h93b2a11a0a843c27(cell=0x00007f6f382865c8) at current.rs:47
    frame #31: 0x00005571b2eb3a5c deadlock`std::thread::local::LocalKey$LT$T$GT$::try_with::h449c927de736c3ab(self=0x00005571b2feacd8, f=closure-0 @ 0x00007f6f38284a78) at local.rs:262
    frame #32: 0x00005571b2eb325d deadlock`std::thread::local::LocalKey$LT$T$GT$::with::hd7a3de31a0f3bb5b(self=0x00005571b2feacd8, f=<unavailable>) at local.rs:239
    frame #33: 0x00005571b2e9e177 deadlock`tokio::runtime::thread_pool::current::set::h4470d759133a25e9(pool=0x00007f6f38284bc8, index=2, f=<unavailable>) at current.rs:29
    frame #34: 0x00005571b2efa613 deadlock`tokio::runtime::thread_pool::worker::Worker::run::hf5a9229692c08b05(self=Worker @ 0x00007f6f38284bc0, blocking_pool=Spawner @ 0x00007f6f38284b50) at worker.rs:145
    frame #35: 0x00005571b2eb253a deadlock`tokio::runtime::thread_pool::Workers::spawn::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h54183d2f9afd5e88 at mod.rs:124
    frame #36: 0x00005571b2ee79e6 deadlock`_$LT$tokio..runtime..blocking..task..BlockingTask$LT$T$GT$$u20$as$u20$core..future..future..Future$GT$::poll::he431127d1f2b019f(self=Pin<&mut tokio::runtime::blocking::task::BlockingTask<closure-0>> @ 0x00007f6f38284c08, _cx=0x00007f6f38284d98) at task.rs:30
    frame #37: 0x00005571b2eeee68 deadlock`tokio::task::core::Core$LT$T$GT$::poll::ha9cae523bf5b35dd(self=0x00005571b3982a40, header=0x00005571b3982a10) at core.rs:128
    frame #38: 0x00005571b2f22938 deadlock`tokio::task::harness::Harness$LT$T$C$S$GT$::poll::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::hd244b0807d34170b at harness.rs:120
    frame #39: 0x00005571b2f09cb8 deadlock`core::ops::function::FnOnce::call_once::he3dc2d61592337ab((null)=closure-0 @ 0x00007f6f38284e50, (null)=<unavailable>) at function.rs:227
    frame #40: 0x00005571b2ef39ce deadlock`_$LT$std..panic..AssertUnwindSafe$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$$LP$$RP$$GT$$GT$::call_once::h43f9b0925ccf6f90(self=AssertUnwindSafe<closure-0> @ 0x00007f6f38284e88, _args=<unavailable>) at panic.rs:317
    frame #41: 0x00005571b2ed3c55 deadlock`std::panicking::try::do_call::hb662d7a89cd6a832(data="\xa0Q(8o\x7f") at panicking.rs:287
    frame #42: 0x00005571b2f75e6a deadlock`__rust_maybe_catch_panic at lib.rs:78
    frame #43: 0x00005571b2ed3787 deadlock`std::panicking::try::h8175f4d1cf0696e4(f=AssertUnwindSafe<closure-0> @ 0x00007f6f38284f78) at panicking.rs:265
    frame #44: 0x00005571b2ef3af5 deadlock`std::panic::catch_unwind::h9c35955090cee57d(f=AssertUnwindSafe<closure-0> @ 0x00007f6f38285008) at panic.rs:396
    frame #45: 0x00005571b2f229f3 deadlock`tokio::task::harness::Harness$LT$T$C$S$GT$::poll::_$u7b$$u7b$closure$u7d$$u7d$::h4586230c80172847((null)=0x00005571b3982a40) at harness.rs:101
    frame #46: 0x00005571b2ed908f deadlock`tokio::loom::std::causal_cell::CausalCell$LT$T$GT$::with_mut::hc910291fd3edb25a(self=0x00005571b3982a40, f=closure-1 @ 0x00007f6f382850a8) at causal_cell.rs:41
    frame #47: 0x00005571b2f21e30 deadlock`tokio::task::harness::Harness$LT$T$C$S$GT$::poll::h14a7d0c8389bbf16(self=Harness<tokio::runtime::blocking::task::BlockingTask<closure-0>, tokio::runtime::blocking::schedule::NoopSchedule> @ 0x00007f6f38285150, executor=&mut FnMut<()> @ 0x00007f6f38285158) at harness.rs:100
    frame #48: 0x00005571b2eefaa8 deadlock`tokio::task::raw::poll::hc3a810f75218ae92(ptr=0x00005571b3982a10, executor=&mut FnMut<()> @ 0x00007f6f382852e0) at raw.rs:162
    frame #49: 0x00005571b2eef8e7 deadlock`tokio::task::raw::RawTask::poll::hc4965144dcaccbf2(self=RawTask @ 0x00007f6f38285320, executor=&mut FnMut<()> @ 0x00007f6f38285328) at raw.rs:113
    frame #50: 0x00005571b2f1736c deadlock`tokio::task::Task$LT$S$GT$::run::hde88a394de623f30(self=Task<tokio::runtime::blocking::schedule::NoopSchedule> @ 0x00007f6f38285360, executor=closure-0 @ 0x00007f6f38285368) at mod.rs:371
    frame #51: 0x00005571b2f04ccd deadlock`tokio::runtime::blocking::pool::run_task::hd53dcc045656c258(f=Task<tokio::runtime::blocking::schedule::NoopSchedule> @ 0x00007f6f382853a8) at pool.rs:368
    frame #52: 0x00005571b2f042e2 deadlock`tokio::runtime::blocking::pool::Inner::run2::h57548d0e6cf6f1ff(self=0x00005571b3983ca0) at pool.rs:284
    frame #53: 0x00005571b2f040a1 deadlock`tokio::runtime::blocking::pool::Inner::run::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h5aabd9c9b07f1bc8 at pool.rs:269
    frame #54: 0x00005571b2e9dfa6 deadlock`tokio::runtime::global::with_state::_$u7b$$u7b$closure$u7d$$u7d$::hfb737014de8337fe(cell=0x00007f6f38286540) at global.rs:100
    frame #55: 0x00005571b2eb4d0b deadlock`std::thread::local::LocalKey$LT$T$GT$::try_with::he8bbbfde4d7bd91e(self=0x00005571b2feacd0, f=closure-0 @ 0x00007f6f38285740) at local.rs:262
    frame #56: 0x00005571b2eb3078 deadlock`std::thread::local::LocalKey$LT$T$GT$::with::hab0004354e063306(self=0x00005571b2feacd0, f=closure-0 @ 0x00007f6f382857d8) at local.rs:239
    frame #57: 0x00005571b2e9ddab deadlock`tokio::runtime::global::with_state::hf5b16a07d731346d(state=State @ 0x00007f6f382857f0, f=closure-0 @ 0x00007f6f38285800) at global.rs:83
    frame #58: 0x00005571b2e9e116 deadlock`tokio::runtime::global::with_thread_pool::h4b2a47496ff40bf8(thread_pool=0x00005571b3983d50, f=closure-0 @ 0x00007f6f38285830) at global.rs:75
    frame #59: 0x00005571b2f0501e deadlock`tokio::runtime::thread_pool::spawner::Spawner::enter::h46cff8db1d056936(self=0x00005571b3983d50, f=closure-0 @ 0x00007f6f38285860) at spawner.rs:44
    frame #60: 0x00005571b2eacd5f deadlock`tokio::runtime::spawner::Spawner::enter::h60f17cb3b42827ca(self=0x00005571b3983d48, f=closure-0 @ 0x00007f6f38285880) at spawner.rs:32
    frame #61: 0x00005571b2f040df deadlock`tokio::runtime::blocking::pool::Inner::run::_$u7b$$u7b$closure$u7d$$u7d$::h30f9ffde2c1a3a10 at pool.rs:269
    frame #62: 0x00005571b2ef31e7 deadlock`tokio::time::clock::Clock::enter::h1fb3c96767342eb4(self=0x00005571b3983ca0, f=closure-0 @ 0x00007f6f382858d8) at clock.rs:30
    frame #63: 0x00005571b2ed97d2 deadlock`tokio::runtime::time::variant::with_default::hcf0dbb8328e5439a(handle=0x00005571b3983d60, clock=0x00005571b3983ca0, f=closure-0 @ 0x00007f6f38285918) at time.rs:43
    frame #64: 0x00005571b2f04058 deadlock`tokio::runtime::blocking::pool::Inner::run::hb8401686d959b775(self=0x00005571b3983ca0) at pool.rs:268
    frame #65: 0x00005571b2f03f61 deadlock`tokio::runtime::blocking::pool::Spawner::spawn_thread::_$u7b$$u7b$closure$u7d$$u7d$::he8e8d568d49a8339 at pool.rs:252
    frame #66: 0x00005571b2ed240b deadlock`std::sys_common::backtrace::__rust_begin_short_backtrace::h56fbfd5cd19ed9b5(f=closure-0 @ 0x00007f6f382859d0) at backtrace.rs:129
    frame #67: 0x00005571b2ef44fe deadlock`std::thread::Builder::spawn_unchecked::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h0f61e7a8a10765b6 at mod.rs:469
    frame #68: 0x00005571b2ef39fe deadlock`_$LT$std..panic..AssertUnwindSafe$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$$LP$$RP$$GT$$GT$::call_once::h4fde050ae54c8cae(self=AssertUnwindSafe<closure-0> @ 0x00007f6f38285a18, _args=<unavailable>) at panic.rs:317
    frame #69: 0x00005571b2ed3b05 deadlock`std::panicking::try::do_call::h451ad55aec6b7ff3(data="\x90<\x98\xb3qU") at panicking.rs:287
    frame #70: 0x00005571b2f75e6a deadlock`__rust_maybe_catch_panic at lib.rs:78
    frame #71: 0x00005571b2ed398a deadlock`std::panicking::try::ha604eb05ec26492b(f=AssertUnwindSafe<closure-0> @ 0x00007f6f38285b00) at panicking.rs:265
    frame #72: 0x00005571b2ef3aae deadlock`std::panic::catch_unwind::h93755e320fffa3bb(f=AssertUnwindSafe<closure-0> @ 0x00007f6f38285ba8) at panic.rs:396
    frame #73: 0x00005571b2ef42ee deadlock`std::thread::Builder::spawn_unchecked::_$u7b$$u7b$closure$u7d$$u7d$::h4a14bf6abf7da27c at mod.rs:468
    frame #74: 0x00005571b2f09824 deadlock`core::ops::function::FnOnce::call_once$u7b$$u7b$vtable.shim$u7d$$u7d$::h2e9adf596166b37b((null)=0x00007f6f38285cc0, (null)=<unavailable>) at function.rs:227
    frame #75: 0x00005571b2f6b7cf deadlock`_$LT$alloc..boxed..Box$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$::call_once::h3534c64212330b0c at boxed.rs:942
    frame #76: 0x00005571b2f75600 deadlock`std::sys::unix::thread::Thread::new::thread_start::h61c012ef60f933c0 [inlined] _$LT$alloc..boxed..Box$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$::call_once::h338c10574a337ece at boxed.rs:942
    frame #77: 0x00005571b2f755f4 deadlock`std::sys::unix::thread::Thread::new::thread_start::h61c012ef60f933c0 [inlined] std::sys_common::thread::start_thread::h761ac6d57710d65d at thread.rs:13
    frame #78: 0x00005571b2f7557a deadlock`std::sys::unix::thread::Thread::new::thread_start::h61c012ef60f933c0 at thread.rs:79
    frame #79: 0x00007f6f38c68ef7 libpthread.so.0`start_thread + 199
    frame #80: 0x00007f6f3898822f libc.so.6`__GI___clone + 63
```

**Comment by dfoxfranke** [comment]
If you jigger the timings, you can get a slightly different backtrace in the hung worker thread:

```
* thread #228, name = 'tokio-runtime-w'
  * frame #0: 0x00007ffff7facee2 libpthread.so.0`__pthread_cond_wait + 274
    frame #1: 0x0000555555650123 deadlock`std::sys::unix::condvar::Condvar::wait::h931978fe7d80a479(self=0x00005555556f3030, mutex=0x00005555556f9de0) at condvar.rs:69
    frame #2: 0x00005555556500d7 deadlock`std::sys_common::condvar::Condvar::wait::h45f6267a90a58bb4(self=0x00005555556f3030, mutex=0x00005555556f9de0) at condvar.rs:41
    frame #3: 0x000055555564b9c1 deadlock`std::sync::condvar::Condvar::wait::h794ed8cfae369a29(self=0x00005555556f2f38, guard=MutexGuard<()> @ 0x00007ffff77c5e00) at condvar.rs:204
    frame #4: 0x00005555555d5b0a deadlock`tokio::runtime::park::Inner::park_condvar::hb8941d5e4205af01(self=0x00005555556f2f20) at park.rs:156
    frame #5: 0x00005555555d54d8 deadlock`tokio::runtime::park::Inner::park::h20609f42a67582ef(self=0x00005555556f2f20) at park.rs:130
    frame #6: 0x00005555555d4fa3 deadlock`_$LT$tokio..runtime..park..Parker$u20$as$u20$tokio..park..Park$GT$::park::hbd86465f394d1bf4(self=0x00005555556f9da0) at park.rs:91
```

The rest of the trace is the same as above, with frame 6 here corresponding to frame 12 above.

I assume this is just a different manifestation of the same bug; it's just a matter of whether the driver happens to be locked when https://github.com/tokio-rs/tokio/blob/34e0a2c04d83877e8bb2e0228818ef0592221e24/tokio/src/runtime/park.rs#L127 is reached. Either way, the end result seems to be the worker thread waiting on a wakeup that never comes.

**Comment by carllerche** [comment]
Thanks for the report. Could you explain why you think it is distinct from #2046?

**Comment by dfoxfranke** [comment]
The hung worker thread has a different backtrace than the one in #2046, and @appaquet's patch from https://github.com/tokio-rs/tokio/compare/master...appaquet:queue-reintrant-deadlock didn't resolve this one.

**Comment by dfoxfranke** [comment]
I just came across this comment and I have a hunch that it describes exactly what's happening here; I'll attempt to confirm when I have the time:

https://github.com/tokio-rs/tokio/blob/7ee554218208a9f3919d840079bd96c6ba881caf/tokio/src/task/state.rs#L444-L453

**Comment by carllerche** [comment]
I don't think it is exactly related to that comment. I have a patch that resolves your repro. I need to figure out why the loom tests did not catch this case though:

```diff
--- a/tokio/src/runtime/thread_pool/worker.rs
+++ b/tokio/src/runtime/thread_pool/worker.rs
@@ -545,15 +545,26 @@ impl GenerationGuard<'_> {
 
         // The worker can only shutdown once there are no further owned tasks.
         while !self.owned().owned_tasks.is_empty() {
+            let mut notify = false;
+            // First, drain all tasks from both the local & global queue.
+            while let Some(task) = self.owned().work_queue.pop_local_first() {
+                notify = true;
+                task.shutdown();
+            }
+
+            if notify {
+                self.worker.slices.notify_all();
+            }
+
+            // Try draining more tasks
+            self.drain_tasks_pending_drop();
+
             // Wait until task that this worker owns are released.
             //
             // `transition_to_parked` is not called as we are not working
             // anymore. When a task is released, the owning worker is unparked
             // directly.
             self.park_mut().park().expect("park failed");
-
-            // Try draining more tasks
-            self.drain_tasks_pending_drop();
         }
     }
```

**Comment by carllerche** [comment]
@dfoxfranke btw, I really appreciate the minimal repro. It made it much easier to debug :+1:

**Comment by carllerche** [comment]
Ok, I believe that this was not detected by loom because it is triggered when receiving an event from the IO driver while the worker is shutting down. The call to `park` visible in the diff above will enter the io driver. If an event is is received, it will result in the task being pushed onto the worker's local queue. The previous assumption was that, once shutdown, tasks would never be pushed onto the local queue again. The fix above drains the local queue every iteration.

Now, the question is how to add a test.

**Comment by carllerche** [comment]
Ideally, we would have a loom test... but the IO driver is disabled during loom tests as it cannot efficiently be mocked out.

I may opt to just copy / paste the repro into the tests dir as we probably should have some full stack tests anyway.

**Comment by carllerche** [comment]
It seems like there is a second issue. I will continue debugging.

**Comment by carllerche** [comment]
The patch above fixes the deadlock when the threaded scheduler is started with 2 core threads. When set to 3, i see another deadlock (probably a separate issue, i am not 100% sure yet).

**Comment by carllerche** [comment]
Actually, the patch above probably adds a bug :D because running the loom tests w/ the patch detects a deadlock...

**Comment by carllerche** [comment]
@dfoxfranke I think #2074 fixes this bug. I am no longer able to reproduce. Are you able to validate?

**Comment by dfoxfranke** [comment]
Confirmed that after applying this patch my test program no longer reproduces the deadlock on my machine. :+1:
